### PR TITLE
Add conversions to/from networkx graphs

### PR DESCRIPTION
**Pull request by uppittu11**
In this PR I added a very simple conversion to and from networkx graphs. At the moment, only site and bond data is transferred when converting. Angles, dihedrals, impropers, subtops, metadata (mixing rules), etc. are ignored making it a lossy conversion to networkx. Let me know how this could be improved.

### CODE REVIEW

**Comment by rsdefever** [comment]
What is the use case for `from_networkx`? I can see the value of  `to_networkx` graph from the topology (as a more abstract representation of the connectivity), but why support the inverse?

**Comment by justinGilmer** [comment]
Personally, the `from_networkx` functionality provides another way to create topologies, from a much more abstract representation of a molecule. Especially if you wanted to iterate through various bonded variations it can sometimes be easier to use a graph-based algorithm.

Although similar things could be done with smiles strings, I want to note.

We can also export a lot of compounds and such as bond graphs, which are a bit more general and can be visualized using a plethora of graph visualization libraries like graphviz. Being able to store these as graph objects and then convert back to a topology can be beneficial.

**Comment by rsdefever** [comment]
> Personally, the `from_networkx` functionality provides another way to create topologies, from a much more abstract representation of a molecule. Especially if you wanted to iterate through various bonded variations it can sometimes be easier to use a graph-based algorithm.

That is a great point, thanks for helping me understand.

File: gmso/external/convert_networkx.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    Notes
+    -----
+    - While a lot of information is lost from converting to a graph object
+    (e.g. metadata, mixing rules, etc.), the graph representation is a
+    useful way to manipulate and extract connectivity information from
+    Topology objects.
+    - The edge has a `connection` attribute, which stores the Bond
+    object it was created from
+    """
+
```
Comment (justinGilmer) [inline gmso/external/convert_networkx.py:34]: I think we should have a validation check here to either ensure it is a `networkx` graph object, or try to coerce into a `networkx` graph

File: gmso/external/convert_networkx.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    -----
+    While a lot of information is lost from converting to a graph object
+    (e.g. metadata, mixing rules, etc.), the graph representation is a
+    useful way to manipulate and extract connectivity information from
+    Topology objects.
+    """
+
+    graph = nx.Graph()
+
+    for b in top.bonds:
```
Comment (justinGilmer) [inline gmso/external/convert_networkx.py:81]: What about the case where a top has no bonds, they would just be edgeless nodes. Would that work?
I dont think a node would be created based on the logic.

Also, if there were two ethanes in the system, would the `Graph` contain two separate graphs within it, or two sets of connected nodes?

File: gmso/tests/test_convert_networkx.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from gmso.core.subtopology import SubTopology as SubTop
+from gmso.core.atom import Atom
+from gmso.external.convert_networkx import from_networkx, to_networkx
+from gmso.tests.base_test import BaseTest
+
+class TestConvertNetworkX(BaseTest):
+    def test_to_networkx_ethane(self, ethane):
+        ethane_to_nx = to_networkx(ethane)
+
+        assert ethane.n_sites == ethane_to_nx.number_of_nodes()
```
Comment (justinGilmer) [inline gmso/tests/test_convert_networkx.py:15]: these are great tests, if we can add in the possibilities above, this is good to go afterwards

**Review by justinGilmer — CHANGES_REQUESTED** [review]
Some small suggestions and a few edge cases (pun _not_ intended)

**Comment by RainierBarrett** [comment]
One slightly hacky way to preserve some more information than just the molecular structure would be to make an unconnected sub-graph with one node, and just make a rule for how that node's weight will hold the metadata etc. that is currently being lost... Any thoughts?

**Review by justinGilmer — CHANGES_REQUESTED** [review]
Looks like the unit tests handle pretty much every case I am concerned with, my only change would be in the from_networkx method. I think we still need a check to ensure that the graph object is a nx graph. Although that could be abstracted out to a separate method like: _valdidate_graph. Although I am not sure that is necessary. Other than that, this is good to merge for me.

File: gmso/external/convert_networkx.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    - The edge has a `connection` attribute, which stores the Bond
+    object it was created from
+    """
+
+    top = Topology()
+
+    node_mapping = dict()
+    for node in graph.nodes:
+        if not isinstance(node, Site):
+            raise GMSOError("Nodes must be gmso.core.Site objects")
```
Comment (umesh-timalsina) [inline gmso/external/convert_networkx.py:40]:
```suggestion
            raise GMSOError("Nodes must be gmso.core.Site objects")
            raise GMSOError("Nodes must be instances of gmso.abc.Site")
```

**Review by umesh-timalsina — APPROVED** [review]

**Review by justinGilmer — APPROVED** [review]

**Comment by CalCraven** [comment]
If you import gmso.external, then you can only see openmm, mbuild, and parmed converters. You have to manually import gmso.external.convert_networkx to get this functionality. Probably worth it to make that converter automatically import when importing gmso.external. 

Also, but just a side note: you have to separately import gmso.external already. Not sure if that is intended or something we want to have available during the initial gmso import.

**Comment by umesh-timalsina** [comment]
```python
(gmso-dev) umesh@isisdell:~$ python
Python 3.6.10 | packaged by conda-forge | (default, Apr 24 2020, 16:44:11) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from gmso.external import convert_networkx
>>> 

```
I am able to import it like the snippet above. @CalCraven, could you point a code snippet to reproduce the issue you mentioned.
